src/compiler/checker.ts

```diff
@@ -45736,6 +45780,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     break;
 
                 case SyntaxKind.ExportKeyword:
+                    if (compilerOptions.verbatimModuleSyntax &&
+                        !(node.flags & NodeFlags.Ambient) &&
+                        node.kind !== SyntaxKind.TypeAliasDeclaration &&
+                        node.kind !== SyntaxKind.InterfaceDeclaration &&
+                        node.parent.kind === SyntaxKind.SourceFile &&
+                        (moduleKind === ModuleKind.CommonJS || getSourceFileOfNode(node).impliedNodeFormat === ModuleKind.CommonJS)
+                    ) {
+                        return grammarErrorOnNode(modifier, Diagnostics.A_top_level_export_modifier_can_only_be_used_on_type_aliases_and_interfaces_in_a_CommonJS_module_when_verbatimModuleSyntax_is_enabled);

```

Comment:
What about `const enum`s?

Reply:
`verbatimModuleSyntax` implies `isolatedModules` which implies `preserveConstEnums`, so yeah, this restriction applies to those too.

Reply:
What about type only namespaces, eg,
```ts
export namespace Debug {
  export type LogLevel = "info" | "debug";
}
```

Reply:
Those would be fine, but it does make this not a _grammar_ error... do you think it’s worth it to change?

Reply:
I don't think anyone particularly cares if something is a grammar error vs a parse error vs a check error - the distinction is largely immaterial to someone reading our compiler output, after all. It's probably just worth fixing because if we don't, I'm sure in a few weeks the feature request will roll in asking us to change it anyway, since namespaces get used to group related types pretty often.

Reply:
Done. Also closed some holes around `export default` I had missed before.